content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) private override document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+Something worth mentioning is that this code _still_ doesn't work. You'll see a compiler error with the following code:
+
+```
+TS2415: Class 'AppComponent' incorrectly extends base class 'BaseComponent'.
+   Types have separate declarations of a private property 'document'.
+```
+
+To solve this, we simply need to make our `BaseComponent`'s `constructor` properties `public` instead of `private`:
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public document: Document) {}
+  ngOnInit() {
+    console.log(document.title);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public override document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+> Remember to keep your `override` property in the `AppComponent` `constructor`, otherwise you'll have errors.
+
+# Why you don't want to extend Angular base classes {#dont-extend-base-classes}
+
+Now that we've learned how to extend base classes in Angular to share lifecycle methods, allow me to flip the script:
+
+**You shouldn't use a base class in Angular**.
+
+<video title="A shocked sock puppet monkey" src="./shocked-monkey-gif.mp4"></video>
+
+Why?
+
+Well, it's often sited by Angular experts that using a base class is brittle and difficult to maintain.
+
+For example, let's say that you have a base component that doesn't use dependency injection, but then suddenly need to add dependency injection. What do you do?
+
+Well, you'd have to refactor every instance that you extended that class.
+
+Similarly, if you add a lifecycle method that you want to overwrite in the future, there can be someheadaches depending in which order you do things in.
+
+There are more than a few ways to write this code in a different, more stable, way.
+
+## Fixing things the right way {#the-fix}
+
+There is a better way to write this code differently today that solve the problems of maintainability a bit better.
+
+Namely, using an `@Injectable` class that's provided on a per-class level enables you to have explicit `setup` and `takedown` functions that you call manually:
+
+```typescript
+@Injectable()
+class WindowSizeService {
+  height = window.innerHeight;
+  width = window.innerWidth;
+ 
+  onResize = () => {
+    this.height = window.innerHeight;
+    this.width = window.innerWidth;
+  }
+
+  addListeners() {
+    window.addEventListener('resize', this.onResize);
+  }
+
+  removeListeners() {
+    window.removeEventListener('resize', this.onResize);
+  }
+}
```

Comment:
*The Angular way* would be to expose an RxJS observable in a service, not regular properties:

```typescript
interface WindowSize {
  readonly height: number;
  readonly width: number;
}

@Injectable()
class WindowSizeService {
  private destroy$ = new Subject<void>();

  size$: Observable<WindowSize>;
  
  constructor(@Inject(DOCUMENT) document: Document) {
    const window = document.defaultView!;
    this.size$ = = fromEvent(this.window, 'resize').pipe(
      debounceTime(50),
      map(() => ({
        height: this.window.innerHeight,
        width: this.window.innerWidth,
      }),
      takeUntil(this.destroy$),
    );
  }
  
  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
}
```

This way, consumers of the service can react to resize events and subscribing to the observable using the `AsyncPipe` or similar will trigger a change detection cycle on the resize events.

Do all of your examples even work in actual Angular apps? 🤔 I'm not sure resize events trigger a change detection cycle unless using a `window:resize` host listener.

Reply:
All of my examples work in actual Angular apps, yeah. I tested.

That said, this _is_ a much better fix. Adding this code sample and attributing it to you.

Thank you for your help